src/rebar_pkg_resource.erl

```diff
@@ -246,19 +246,35 @@ serve_from_cache(TmpDir, CachePath, Pkg) ->
       Tarball :: binary(),
       Package :: package(),
       Res :: ok | {error,_} | {bad_registry_checksum, integer(), integer()}.
-serve_from_memory(TmpDir, Binary, {pkg, _Name, _Vsn, Hash, _RepoConfig}) ->
+serve_from_memory(TmpDir, Binary, {pkg, _Name, _Vsn, OldHash, Hash, _RepoConfig}) ->
     RegistryChecksum = list_to_integer(binary_to_list(Hash), 16),
+    OldRegistryChecksum =  maybe_old_registry_checksum(OldHash),
     case r3_hex_tarball:unpack(Binary, TmpDir) of
-        {ok, #{checksum := <<Checksum:256/big-unsigned>>}} when RegistryChecksum =/= Checksum ->
-            ?DEBUG("Expected hash ~64.16.0B does not match checksum of fetched package ~64.16.0B",
-                   [RegistryChecksum, Checksum]),
-            {bad_registry_checksum, RegistryChecksum, Checksum};
-        {ok, #{checksum := <<RegistryChecksum:256/big-unsigned>>}} ->
+        {ok, #{outer_checksum := <<Checksum:256/big-unsigned>>} = Res} when RegistryChecksum =/= Checksum ->
+            #{inner_checksum := <<OldChecksum:256/big-unsigned>>} = Res,
+            %% Not triggerable in tests, but code feels logically wrong without it since inner checksums are not hard
+            %% deprecated. This logic should be removed when inner checksums do become hard deprecated and/or no longer
+            %% supported by rebar3.
+            case OldRegistryChecksum == OldChecksum of
+                true ->
+                    ?DEBUG("Expected hash ~64.16.0B does not match outer checksum of fetched package ~64.16.0B, but
+                           matches inner checksum ~64.16.0B",
+                        [RegistryChecksum, Checksum, OldChecksum]),
+                    ok;
```

Comment:
I think we should, at least for the RC-1, crash here regardless, but keep the new logging message. This case could technically be the attack the outer-checksum is planning to protect against.

Reply:
sounds good